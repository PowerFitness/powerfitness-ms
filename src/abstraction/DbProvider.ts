import { DbConfig } from '../config'
import { createConnection, Connection, MysqlError } from 'mysql';

export class DbProvider {
	dbConfig: DbConfig;
	connection: Connection | null = null;
	constructor(dbConfig: DbConfig) {
		this.dbConfig = dbConfig;
	}
	establishConnection(): void {
		if(!this.connection) {
			this.connection = createConnection(this.dbConfig)
		}
		return;
	}
	execute<T>(query: string): Promise<T> {
		this.establishConnection();
		return new Promise((resolve, reject) => {
			(this.connection as Connection).query(query, (error: MysqlError, results: T) => {
				if (error) {
					reject(error)
				} else {
					resolve(results)
				}
				(this.connection as Connection).end()
			})
		}) as unknown as Promise<T>;
	}

	beginTransaction(): Promise<void> {
		this.establishConnection();
		return new Promise((resolve, reject) =>
			(this.connection as Connection).beginTransaction((err: MysqlError) => err ? reject(err) : resolve()));
	}

	commit(): Promise<void> {
		this.establishConnection();
		return new Promise((resolve, reject) => (this.connection as Connection).commit((err: MysqlError) => {
			if (err) {
				this.rollback()
					.then(() => reject(err))
					.catch((rollbackError: MysqlError) =>
						reject(new Error(`Rollback Failed: ${err.message} / ${rollbackError.message}`)))
					.finally(() =>
						(this.connection as Connection).end())
			} else {
				resolve();
				(this.connection as Connection).end()
			}
		}));
	}

	rollback(): Promise<void> {
		return new Promise((resolve, reject) =>
			(this.connection as Connection).rollback((err: MysqlError) => err ? reject(err) : resolve()));
	}
}

export default DbProvider;
