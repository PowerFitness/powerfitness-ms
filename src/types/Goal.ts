export interface Goal {
    id: number;
    planId: number;
    type: string;
    name: string;
    unit: string;
    value: number;
    createdDate: string;
    updatedDate: string;
}
