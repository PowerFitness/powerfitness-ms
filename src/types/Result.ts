export interface Result {
	id: number;
	userUniqueId: string;
	date: string;
	type: string;
	subtype: string;
	name: string;
	unit: string;
	value: number;
	createdDate: string;
	lastUpdatedDate: string;
}
