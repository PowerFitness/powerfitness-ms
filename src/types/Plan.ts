import { Goal } from './Goal';

export interface Plan {
	id: number;
	userUniqueId: string;
	motivationStatement: string;
	createdDate: string;
	lastUpdatedDate: string;
    goals?: Array<Goal>
}
