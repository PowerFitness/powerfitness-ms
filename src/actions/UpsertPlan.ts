import { Action } from '../abstraction/Action';
import DbProvider from '../abstraction/DbProvider';
import { OkPacket, escape } from 'mysql';
import { Plan } from '../types/Plan';

export class UpsertPlan extends Action<OkPacket> {
	plan: Plan;
	constructor(dbProvider: DbProvider, plan: Plan) {
		super(dbProvider);
		this.plan = plan;
	}
	buildQuery(): string {
		const { id, userUniqueId, motivationStatement } = this.plan;
		return `insert into plan 
					(id, userUniqueId, motivationStatement, createdDate, lastUpdatedDate)
				values 
					(
						${id ? escape(id) : null}, 
						${escape(userUniqueId)}, 
						${escape(motivationStatement)}, 
						${'CURRENT_TIMESTAMP()'}, 
						${'CURRENT_TIMESTAMP()'}
					)
				on duplicate key update
					motivationStatement = ${escape(motivationStatement)}, lastUpdatedDate = ${'CURRENT_TIMESTAMP()'}`;
	}
}

export default UpsertPlan;
