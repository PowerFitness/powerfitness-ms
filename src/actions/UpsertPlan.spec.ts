import DbProvider from '../abstraction/DbProvider';
import { Plan } from '../types/Plan';
import UpsertPlan from './UpsertPlan';
import { mock } from 'jest-mock-extended';

describe('UpsertPlan', () => {
	test('buildQuery', () => {
		const plan: Plan = {
			id: 1,
			userUniqueId: 'uid',
			motivationStatement: 'To be healthy',
			createdDate: 'date',
			lastUpdatedDate: 'date'
		};
		const dbProvider: DbProvider = mock<DbProvider>();
		const action: UpsertPlan = new UpsertPlan(dbProvider, plan);
		expect(action.buildQuery()).toMatchSnapshot();
	})
})
