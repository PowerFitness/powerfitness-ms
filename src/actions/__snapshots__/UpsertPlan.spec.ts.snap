// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpsertPlan buildQuery 1`] = `
"insert into plan 
					(id, userUniqueId, motivationStatement, createdDate, lastUpdatedDate)
				values 
					(
						1, 
						'uid', 
						'To be healthy', 
						CURRENT_TIMESTAMP(), 
						CURRENT_TIMESTAMP()
					)
				on duplicate key update
					motivationStatement = 'To be healthy', lastUpdatedDate = CURRENT_TIMESTAMP()"
`;

exports[`UpsertPlan buildQuery without id 1`] = `
"insert into plan 
					(id, userUniqueId, motivationStatement, createdDate, lastUpdatedDate)
				values 
					(
						null, 
						'uid', 
						'To be healthy', 
						CURRENT_TIMESTAMP(), 
						CURRENT_TIMESTAMP()
					)
				on duplicate key update
					motivationStatement = 'To be healthy', lastUpdatedDate = CURRENT_TIMESTAMP()"
`;
